Karma-Bond/Contracts/Karma-CustomBond/src/main/java/dao/karma/custombond/KarmaCustomBond.java

```diff
@@ -697,10 +700,100 @@ public BigInteger bondPrice() {
         if (price.compareTo(terms.minimumPrice) < 0) {
             price = terms.minimumPrice;
         }
+        // check if max discount is greater than 0 and increase price to fit the capped discount
+        // NOTE: if minimumPrice is set in the terms capped discount is not applied!
+        else if (terms.maxDiscount.compareTo(ZERO) > 0) {
+            BigInteger bondDiscount = _currentBondDiscount(price);
+            BigInteger maxDiscount = terms.maxDiscount;
+
+            // if bond discount is greater than max discount, increase bond price to fit the max discount
+            if (bondDiscount.compareTo(maxDiscount) > 0) {
+                BigInteger newTrueBondPrice = this.payoutTokenMarketPriceUSD().subtract(maxDiscount.multiply(payoutTokenMarketPriceUSD())
+                                .divide(MathUtils.pow10(3))).multiply(MathUtils.pow10(7)).divide(this.principalTokenMarketPriceUSD());
+                BigInteger newBondPrice = newTrueBondPrice.subtract(newTrueBondPrice.multiply(currentKarmaFee()).divide(MathUtils.pow10(6)));
+
+                // only apply new bond price if it is higher than the old one, this should mitigate oracle risk
+                // by defaulting to the un-capped bond price
+                if (newBondPrice.compareTo(price) > 0) {
+                    price = newBondPrice;
+                }
+            }
+        }
 
         return price;
     }
 
+    /**
+     *  Calculate bond discount using market USD prices of principal and payout token
+     *  @return BigInteger - Discount in percentages. Divide by 1e7 and multiply by 100% to convert in percentages
+     *                       on client side.
+     */
+    @External(readonly = true)
+    public BigInteger currentBondDiscount() {
+        BigInteger bondPriceUSD = this.bondPriceUSD();
+        BigInteger payoutTokenMarketPriceUSD = payoutTokenMarketPriceUSD();
+
+        // discount = (payout token market price USD - bond price USD) / payout token market price USD
+        // NOTE: result is in 1e7 decimal precision
+        return (payoutTokenMarketPriceUSD.subtract(bondPriceUSD)).divide(payoutTokenMarketPriceUSD.divide(MathUtils.pow10(7)));
+    }
+
+    /**
+     *  Calculate current bond discount
+     *  @param bondPrice - Current bond price
+     *  @return BigInteger - Discount in percentages. Divide by 1e7 and multiply by 100% to convert in percentages
+     *                       on client side.
+     */
+    @External(readonly = true)
+    public BigInteger _currentBondDiscount(BigInteger bondPrice) {

```

Comment:
We may avoid code duplication between `currentBondDiscount`  and  `_currentBondDiscount` and `bondPriceUSD` and `_bondPriceUSD` like this : 

```java
@External(readonly = true)
public BigInteger currentBondDiscount(
  @Optional BigInteger bondPrice // <---- note this optional parameter
) {
  BigInteger bondPriceUSD = this.bondPriceUSD(bondPrice);
  // [...] same as before
}


@External(readonly = true)
public BigInteger bondPriceUSD(BigInteger bondPrice) {
  // NOTE: result is in 1e18 decimal precision
  return this.trueBondPrice(bondPrice).multiply(this.principalTokenMarketPriceUSD()).divide(MathUtils.pow10(7));
}


@External(readonly = true)
public BigInteger trueBondPrice(
  @Optional BigInteger bondPrice // <---- note this optional parameter
) {
  // truePrice = `bondPrice()` + (`bondPrice()` * `currentKarmaFee()` / 10**TRUE_BOND_PRICE_PRECISION)
  if (bondPrice == null) {
    // recursion only if no parameter given
    bondPrice = this.bondPrice()  
  }
  return bondPrice.add(bondPrice.multiply(currentKarmaFee()).divide(MathUtils.pow10(TRUE_BOND_PRICE_PRECISION)));
}
```

That way there is no duplicated code, and the code keeps working the same than before